### PR TITLE
Fix QOS Test (and add cleanup)

### PR DESCRIPTION
**Pull request by garrettmoon**
- Remove lower quality images from cache if a higher quality one is to be downloaded
- Document weirdness around QOS and caching
- Fix QOS test

### CODE REVIEW

File: Example/PINRemoteImage Tests/PINRemoteImage_Tests.m [hunk 1]
```diff
@@ -204,6 +204,16 @@ - (void)drawImage:(UIImage *)image
     UIGraphicsEndImageContext();
 }
 
+- (void)waitForImageWithURLToBeCached:(NSURL *)URL
+{
+    for (NSUInteger idx = 0; idx < 100; idx++) {
+        if ([[self.imageManager cache] objectForKey:[self.imageManager cacheKeyForURL:URL processorKey:nil]] != nil) {
```
Comment (jparise) [inline Example/PINRemoteImage Tests/PINRemoteImage_Tests.m]: You could call `-cacheForKey:processorKey:` outside of the loop.

File: Example/PINRemoteImage Tests/PINRemoteImage_Tests.m [hunk 2]
```diff
@@ -594,6 +605,15 @@ - (void)testQOS
      }];
     dispatch_semaphore_wait(semaphore, [self timeout]);
     
+    //small image should have been removed from cache
+    for (NSUInteger idx = 0; idx < 100; idx++) {
```
Comment (jparise) [inline Example/PINRemoteImage Tests/PINRemoteImage_Tests.m]: Can you use `-waitForImageWithURLToBeCached` here?

File: Pod/Classes/PINRemoteImageManager.m [hunk 3]
```diff
@@ -1165,23 +1176,35 @@ - (NSUUID *)downloadImageWithURLs:(NSArray *)urls
...
                             processorKey:nil
                                processor:nil
                                 progress:progress
-                              completion:completion
+                              completion:^(PINRemoteImageManagerResult *result) {
+                                  typeof(self) strongSelf = weakSelf;
+                                  //clean out any lower quality images from the cache
+                                  for (NSInteger idx = downloadIdx - 1; idx >= 0; idx--) {
+                                      [[strongSelf cache] removeObjectForKey:[strongSelf cacheKeyForURL:[urls objectAtIndex:idx] processorKey:nil]];
+                                  }
```
Comment (jparise) [inline Pod/Classes/PINRemoteImageManager.m]: This looks fine but makes me think we should consider a `-removeObjectsForKeys:(NSArray *)` method in PINCache that can do this work in a single locked batch operation.

Comment (garrettmoon) [inline Pod/Classes/PINRemoteImageManager.m]: That's a good idea.

File: Example/PINRemoteImage Tests/PINRemoteImage_Tests.m (hunk 2 above)
Comment (garrettmoon) [inline Example/PINRemoteImage Tests/PINRemoteImage_Tests.m]: No, this is checking for the opposite, that the image is *not* in the cache.

**Pull request merged**